{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bff91b09",
   "metadata": {},
   "source": [
    "# Exercises Week 11-12: Graded exercise session (part B)\n",
    "\n",
    "**Course**: [Life Sciences engineering: genome to function](https://go.epfl.ch/BIO-411) (BIO-411)\n",
    "\n",
    "**Professors**:  _Gönczy Pierre_, _Naef Felix_, _McCabe Brian Donal_\n",
    "\n",
    "SSV, MA, 2023-2024"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "867ce50e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.decomposition import PCA\n",
    "import plotly.express as px\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.preprocessing import scale\n",
    "\n",
    "from scipy.integrate import odeint\n",
    "import matplotlib.pyplot as plt\n",
    "from ipywidgets import interact, fixed\n",
    "from scipy.stats import beta\n",
    "import math"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e04f778e",
   "metadata": {},
   "source": [
    "**Provide answers in the form of code, figures (if relevant) and short descriptions (in markdown cells) in those notebooks. Submit your notebook to Moodle, please make sure to execute every cell.**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4dcb272",
   "metadata": {},
   "source": [
    "### Exercise 1: Simulation of circadian gene regulation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecaa0b5a",
   "metadata": {},
   "source": [
    "In this exercise, we revisit and extend the concepts from Week 9, focusing on the transcription and splicing of a pre-mRNA to produce mRNA. We describe this process using a two-dimensional Ordinary Differential Equation (ODE), similar to our previous approach but with a key difference: we now incorporate **time-dependent** transcription and degradation rates to model circadian rhythms.\n",
    "\n",
    "The system is defined by the following ODEs:\n",
    "\n",
    "\\begin{array}{ll}\n",
    "\\frac{dP}{dt} = s(t) - \\rho P \\\\  \n",
    "\\\\\n",
    "\\frac{dM}{dt} = \\rho P - k(t) M \\\\\n",
    "\\end{array}\n",
    "\n",
    "where  $P$ and $M$ represent the concentrations of pre-mRNA and mRNA, respectively. We will consider time-dependent (circadian) transcription and degradation rates. The frequency is defined by $\\omega=\\frac{2\\pi}{T}$ with $T=24h$.  \n",
    "\n",
    "$s(t)$ is the transcription rate, which is now taken either as constant $s(t)=s_{0}$ or as a rhythmic function of time:  \n",
    "\n",
    "\\begin{equation*}\n",
    "s(t) = s_0(1 + \\epsilon_s \\cos(\\omega t)) \n",
    "\\end{equation*}  \n",
    "with relative amplitude $0\\leq \\epsilon_s\\leq 1$. Note that the peak of $s(t)$ is at $t=0$.  \n",
    " \n",
    "Similarly, we will consider the degradation rate $k(t)$ either as constant $k(t)= k_0$ or a rhythmic function of time:  \n",
    "\\begin{equation*}\n",
    "k(t)=k_0(1 + \\epsilon_k \\cos(\\omega (t-t_k))\n",
    "\\end{equation*}  \n",
    "with $0\\leq\\epsilon_k\\leq 1$, and the maximum degradation rate at $t=t_k$.\n",
    "\n",
    "$\\rho$ represents the constant splicing rate of pre-mRNA $P$ into mRNA $M$."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebecf369",
   "metadata": {},
   "source": [
    "### Question 1\n",
    "1. Adapt the code from Week 9 exercises to simulate this system with rhythmic transcription and degradation rates. Modify the interactive widget such that you can vary the new parameters and plot $P(t)$ and $M(t)$.  \n",
    "\n",
    "2. Keep the degradation rate constant (set $\\epsilon_k$=0) and vary $k_0$ and $\\epsilon_s$.\n",
    "    * How does the phase delay between $M(t)$ and $P(t)$ changes with different values of $k_0$?\n",
    "    * How does the amplitude (use the ratio between max and min) of $M(t)$ changes with different values of $k_0$? \n",
    "    \n",
    "    Make sure you span a relevant range for $k_0^{-1}$ ( *i.e.* from 10 minutes to several hours).  \n",
    "\n",
    "3. Now introduce rhythmic degradation by varying $\\epsilon_k$. \n",
    "    * Show that you can obtain a rhyhtmic mRNA accumulatiokn when the synthesis is constant ($\\epsilon_s=0$). \n",
    "    * Explore and discuss scenarios where the introduction of rhythmic degradation leads to dynamics that cannot be achieved with a constant half-life.  \n",
    "    \n",
    "    See also [Wang et al., 2018](https://www.pnas.org/doi/10.1073/pnas.1715225115)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def production_decay_ode(y, time, rho, s0, es, k0, ek, tk ):\n",
    "\n",
    "    \"\"\"\n",
    "    Defines the ODEs for the production and decay of pre-mRNA and mRNA.\n",
    "    P: pre-mRNA concentration\n",
    "    M: mRNA concentration\n",
    "    st: transcription rate (time dependent)\n",
    "    kt: mRNA degradation rate (time dependent)\n",
    "    rho: splicing rate\n",
    "    s0: transcription constant\n",
    "    es: relative amplitude of the transcription rate oscillation\n",
    "    ek: relative amplitude of the mRNA degradation rate oscillation\n",
    "    k0: degradation constant\n",
    "    time: time point (in minutes)\n",
    "    tk: max degrade rate\n",
    "\n",
    "    #kp: pre-mRNA degradation rate\n",
    "    #km: mRNA degradation rate\n",
    "    #tau: protein synthesis rate\n",
    "    #kg: protein degradation rate\n",
    "    \"\"\"\n",
    "    omega = 2*np.pi/(24*60)\n",
    "    P, M = y\n",
    "    st = s0*(1+es*math.cos(omega*time))\n",
    "    kt = k0*(1+ek*math.cos(omega*(time-tk)))\n",
    "    dPdt = st - rho*P\n",
    "    dMdt = rho*P - kt*M\n",
    "    return [dPdt, dMdt] \n",
    "\n",
    "def simulation(ode_function, parameters, initial_conditions = None, time = None):\n",
    "\n",
    "    \"\"\"\n",
    "    Simulate the ODEs for the production and decay of pre-mRNA and mRNA.\n",
    "    parameters: parameters for the ODEs\n",
    "    initial_conditions: initial conditions for the ODEs\n",
    "    time: time intervals for the simulation\n",
    "   \n",
    "    s = parameters['s']\n",
    "    kp = parameters['kp']\n",
    "    km = parameters['km']\n",
    "    rho = parameters['rho']\n",
    "    tau = parameters['tau']\n",
    "    kg = parameters['kg']\n",
    "    \"\"\"\n",
    "    \n",
    "    \n",
    "    \n",
    "    rho= parameters['rho']\n",
    "    s0= parameters['s0']\n",
    "    es= parameters['es']\n",
    "    k0= parameters['k0']\n",
    "    ek= parameters['ek']\n",
    "    #time= parameters['time']\n",
    "    tk= parameters['tk']\n",
    "    \n",
    "    #omega = 2*np.pi/(24*60)\n",
    "    #st = s0*(1+es*math.cos(omega*time))\n",
    "    #kt = k0*(1+ek*math.cos(omega*(time-tk)))\n",
    "    # Placeholder for adding steady-state solutions\n",
    "\n",
    "    P_ss = s0 / rho #Add your steady-state solution for P here \n",
    "    M_ss = s0 / k0 #Add you steady-state solution for M here\n",
    "    \n",
    "    #results = odeint(ode_function, initial_conditions, time, args=(rho, s0, es, k0, ek, tk))\n",
    "    results = odeint(ode_function, initial_conditions, time, args=(rho, s0, es, k0, ek, tk))\n",
    "\n",
    "    plt.figure(figsize=(10, 6))\n",
    "    plt.ylim([0,150])\n",
    "\n",
    "    # Uncomment the lines below after adding your steady-state solutions\n",
    "    plt.axhline(y=P_ss, color='blue', linestyle='--', label='P_ss')  \n",
    "    plt.axhline(y=M_ss, color='orange', linestyle='--', label='M_ss')\n",
    "\n",
    "    plt.title('Simulation of pre-mRNA and mRNA Production-Decay Dynamics')\n",
    "    plt.plot(time, results[:, 0], label = 'P(t)', linewidth=2)\n",
    "    plt.plot(time, results[:, 1], label = 'M(t)', linewidth=2)\n",
    "    plt.legend()\n",
    "\n",
    "    plt.xlabel(\"Time [min]\")\n",
    "    plt.ylabel(\"Concentration\")\n",
    "    plt.legend(loc='upper left')\n",
    "    plt.show()\n",
    "    \n",
    "def animate(initial_conditions, time, rho=0.4, s0=5, es=1, k0=0.05, ek=1, tk=0.5 ):\n",
    "    \"\"\"\n",
    "    Runs the simulation with an interactive widgets for parameter exploration.\n",
    "    \"\"\"\n",
    "    #args=(rho, s0, es, k0, ek, time, tk )\n",
    "    parameters = {'rho':rho, 's0':s0, 'es':es, 'k0':k0, 'ek':ek, 'tk':tk}\n",
    "    simulation(production_decay_ode, parameters, initial_conditions, time)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3baafac4cf7a4c3aad96f645ede978fe",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "interactive(children=(FloatSlider(value=0.4, description='rho', max=1.0), FloatSlider(value=5.0, description='…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Define initial conditions and time intervals for the simulation\n",
    "initial_conditions = [0, 0] #[P, M, G]\n",
    "time = np.linspace(0, 500, 10001)\n",
    "\n",
    "interact(animate, \n",
    "         initial_conditions=fixed(initial_conditions), \n",
    "         time=fixed(time), \n",
    "        rho=(0, 1, 0.1),\n",
    "        s0=(0, 10, 0.1),\n",
    "        es=(0, 1, 0.1),\n",
    "        k0=(1/600,1/10,1/1000),\n",
    "        ek=(0, 1, 0.1),\n",
    "        tk=(0, 1, 0.1));"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94467283",
   "metadata": {},
   "source": [
    "### Exercise 2: Circadian (post-)transcriptional regulation of gene expression in mouse liver"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d4c6d92",
   "metadata": {},
   "source": [
    "In this exercise, we will analyse published RNA-seq data from mouse liver under an *ad libitum* feeding regimen and in presence of a 12-hour light-dark cycle ([Atger et al., 2015](https://www.pnas.org/doi/abs/10.1073/pnas.1515308112)). The samples were harvested every 2 hours in four replicates, RNA was extracted and sequenced. Similarly to Week 9 exercises on RNA-seq data, gene expression was quantified at the intron and exon levels. Note that data are $log_2$ transformed and normalized (RPKM)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcf58338",
   "metadata": {},
   "source": [
    "### Question 1\n",
    "1. Modify the code from Week 9 exercises to perform PCA using the 14 following genes (core clock genes and selected clock output genes):  \n",
    " ``['Arntl', 'Npas2', 'Clock', 'Cry1', 'Cry2', 'Per1', 'Per2', 'Per3', 'Nr1d1', 'Nr1d2', 'Rora', 'Rorc', 'Tef', 'Dbp']``.  \n",
    " First start with only the exons, then only the introns, and eventually both. Describe your observations.  \n",
    "\n",
    "2. Use the *return_amp_phase_pv* function on the $log_2$ data to assess and characterize rhythmicity genome-wide.\n",
    "    * Select genes based on p-value ($p<0.01) and amplitude (log2 fold-change>0.5) thresholds for further analysis.\n",
    "    * Provide histograms of the peak times, and amplitudes ($log_2$ peak-to-trough) for the selected genes.\n",
    "    * Perform this analysis separately for the exons and introns.\n",
    "    * Examine whether you observe patterns typically referred to as the 'morning' and 'evening' waves of gene expression.\n",
    "    \n",
    "3. Determine which genes show the highest amplitude rhythmic patterns. Are there any common themes or characteristics among these genes?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d725b9ef",
   "metadata": {},
   "source": [
    "#### Load the RNA-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a51a74b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>condition</th>\n",
       "      <th colspan=\"21\" halign=\"left\">WT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feeding</th>\n",
       "      <th colspan=\"21\" halign=\"left\">AL</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature</th>\n",
       "      <th colspan=\"10\" halign=\"left\">Intron</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"10\" halign=\"left\">RFP</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>time</th>\n",
       "      <th>00</th>\n",
       "      <th>02</th>\n",
       "      <th>04</th>\n",
       "      <th>06</th>\n",
       "      <th>08</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>14</th>\n",
       "      <th>16</th>\n",
       "      <th>18</th>\n",
       "      <th>...</th>\n",
       "      <th>04</th>\n",
       "      <th>06</th>\n",
       "      <th>08</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>14</th>\n",
       "      <th>16</th>\n",
       "      <th>18</th>\n",
       "      <th>20</th>\n",
       "      <th>22</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>replicate</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>...</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gnai3</th>\n",
       "      <td>-1.564402</td>\n",
       "      <td>-1.536626</td>\n",
       "      <td>-1.931707</td>\n",
       "      <td>-1.458354</td>\n",
       "      <td>-1.575970</td>\n",
       "      <td>-1.654038</td>\n",
       "      <td>-1.550707</td>\n",
       "      <td>-1.564596</td>\n",
       "      <td>-1.211372</td>\n",
       "      <td>-1.654187</td>\n",
       "      <td>...</td>\n",
       "      <td>4.739938</td>\n",
       "      <td>4.648414</td>\n",
       "      <td>4.576937</td>\n",
       "      <td>4.891934</td>\n",
       "      <td>5.284247</td>\n",
       "      <td>5.502545</td>\n",
       "      <td>5.115915</td>\n",
       "      <td>5.402904</td>\n",
       "      <td>5.119890</td>\n",
       "      <td>4.873039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cdc45</th>\n",
       "      <td>-3.838233</td>\n",
       "      <td>-4.132913</td>\n",
       "      <td>-4.023130</td>\n",
       "      <td>-3.782926</td>\n",
       "      <td>-3.971205</td>\n",
       "      <td>-3.898270</td>\n",
       "      <td>-3.837377</td>\n",
       "      <td>-3.907227</td>\n",
       "      <td>-4.002231</td>\n",
       "      <td>-3.837719</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.783427</td>\n",
       "      <td>-1.045727</td>\n",
       "      <td>-1.868940</td>\n",
       "      <td>-1.616297</td>\n",
       "      <td>1.552717</td>\n",
       "      <td>0.336330</td>\n",
       "      <td>-1.370050</td>\n",
       "      <td>-0.308212</td>\n",
       "      <td>-1.218384</td>\n",
       "      <td>-0.541827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Apoh</th>\n",
       "      <td>-1.075680</td>\n",
       "      <td>-1.540220</td>\n",
       "      <td>-1.569142</td>\n",
       "      <td>-1.652602</td>\n",
       "      <td>-1.786800</td>\n",
       "      <td>-1.539642</td>\n",
       "      <td>-1.434320</td>\n",
       "      <td>-1.655522</td>\n",
       "      <td>-1.499171</td>\n",
       "      <td>-1.411709</td>\n",
       "      <td>...</td>\n",
       "      <td>10.657651</td>\n",
       "      <td>10.460193</td>\n",
       "      <td>10.623698</td>\n",
       "      <td>10.381588</td>\n",
       "      <td>10.146668</td>\n",
       "      <td>10.232355</td>\n",
       "      <td>10.631922</td>\n",
       "      <td>10.468625</td>\n",
       "      <td>10.704931</td>\n",
       "      <td>10.591210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Narf</th>\n",
       "      <td>-1.198793</td>\n",
       "      <td>-0.937221</td>\n",
       "      <td>-0.921084</td>\n",
       "      <td>-0.331866</td>\n",
       "      <td>-0.123671</td>\n",
       "      <td>-0.186613</td>\n",
       "      <td>-0.632810</td>\n",
       "      <td>-0.963087</td>\n",
       "      <td>-0.999763</td>\n",
       "      <td>-1.320254</td>\n",
       "      <td>...</td>\n",
       "      <td>1.788997</td>\n",
       "      <td>2.660656</td>\n",
       "      <td>2.890531</td>\n",
       "      <td>3.379632</td>\n",
       "      <td>2.924193</td>\n",
       "      <td>2.472548</td>\n",
       "      <td>2.686118</td>\n",
       "      <td>2.390108</td>\n",
       "      <td>1.941836</td>\n",
       "      <td>2.128767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cav2</th>\n",
       "      <td>-3.046334</td>\n",
       "      <td>-3.741371</td>\n",
       "      <td>-3.364087</td>\n",
       "      <td>-3.063080</td>\n",
       "      <td>-3.374724</td>\n",
       "      <td>-2.992739</td>\n",
       "      <td>-2.985271</td>\n",
       "      <td>-2.991166</td>\n",
       "      <td>-2.954387</td>\n",
       "      <td>-3.052015</td>\n",
       "      <td>...</td>\n",
       "      <td>0.759856</td>\n",
       "      <td>0.403964</td>\n",
       "      <td>0.581452</td>\n",
       "      <td>0.716145</td>\n",
       "      <td>0.742948</td>\n",
       "      <td>0.385111</td>\n",
       "      <td>0.282426</td>\n",
       "      <td>0.624375</td>\n",
       "      <td>0.806800</td>\n",
       "      <td>0.524357</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 144 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "condition          WT                                                    \\\n",
       "feeding            AL                                                     \n",
       "feature        Intron                                                     \n",
       "time               00        02        04        06        08        10   \n",
       "replicate           A         A         A         A         A         A   \n",
       "Gene_Symbol                                                               \n",
       "Gnai3       -1.564402 -1.536626 -1.931707 -1.458354 -1.575970 -1.654038   \n",
       "Cdc45       -3.838233 -4.132913 -4.023130 -3.782926 -3.971205 -3.898270   \n",
       "Apoh        -1.075680 -1.540220 -1.569142 -1.652602 -1.786800 -1.539642   \n",
       "Narf        -1.198793 -0.937221 -0.921084 -0.331866 -0.123671 -0.186613   \n",
       "Cav2        -3.046334 -3.741371 -3.364087 -3.063080 -3.374724 -2.992739   \n",
       "\n",
       "condition                                            ...             \\\n",
       "feeding                                              ...              \n",
       "feature                                              ...        RFP   \n",
       "time               12        14        16        18  ...         04   \n",
       "replicate           A         A         A         A  ...          D   \n",
       "Gene_Symbol                                          ...              \n",
       "Gnai3       -1.550707 -1.564596 -1.211372 -1.654187  ...   4.739938   \n",
       "Cdc45       -3.837377 -3.907227 -4.002231 -3.837719  ...  -1.783427   \n",
       "Apoh        -1.434320 -1.655522 -1.499171 -1.411709  ...  10.657651   \n",
       "Narf        -0.632810 -0.963087 -0.999763 -1.320254  ...   1.788997   \n",
       "Cav2        -2.985271 -2.991166 -2.954387 -3.052015  ...   0.759856   \n",
       "\n",
       "condition                                                                      \\\n",
       "feeding                                                                         \n",
       "feature                                                                         \n",
       "time                06         08         10         12         14         16   \n",
       "replicate            D          D          D          D          D          D   \n",
       "Gene_Symbol                                                                     \n",
       "Gnai3         4.648414   4.576937   4.891934   5.284247   5.502545   5.115915   \n",
       "Cdc45        -1.045727  -1.868940  -1.616297   1.552717   0.336330  -1.370050   \n",
       "Apoh         10.460193  10.623698  10.381588  10.146668  10.232355  10.631922   \n",
       "Narf          2.660656   2.890531   3.379632   2.924193   2.472548   2.686118   \n",
       "Cav2          0.403964   0.581452   0.716145   0.742948   0.385111   0.282426   \n",
       "\n",
       "condition                                     \n",
       "feeding                                       \n",
       "feature                                       \n",
       "time                18         20         22  \n",
       "replicate            D          D          D  \n",
       "Gene_Symbol                                   \n",
       "Gnai3         5.402904   5.119890   4.873039  \n",
       "Cdc45        -0.308212  -1.218384  -0.541827  \n",
       "Apoh         10.468625  10.704931  10.591210  \n",
       "Narf          2.390108   1.941836   2.128767  \n",
       "Cav2          0.624375   0.806800   0.524357  \n",
       "\n",
       "[5 rows x 144 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dat = pd.read_csv(\"./GSE73554_WT_AL_Intron_Exon_RFP.txt\",sep='\\t')\n",
    "dat.index = dat['Gene_Symbol']\n",
    "dat = dat.drop(['Gene_Symbol','Gene_Ensembl'], axis=1)\n",
    "dat.columns = dat.columns.str.split('_', expand = True)\n",
    "dat.columns.names = ['condition','feeding','feature','time','replicate']\n",
    "\n",
    "dat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def subselect(data, feature, replicate, time, condition, feeding):\n",
    "\n",
    "    \"\"\"\n",
    "    Subselects samples from the full dataset based on the specified features, conditions, \n",
    "    replicates, and time points.\n",
    "\n",
    "    Parameters:\n",
    "    data: The full gene expression dataset.\n",
    "    feature: The features to filter by (e.g., ['intron', 'exon']).\n",
    "    replicate: The replicates to include.\n",
    "    time: The time points to include.\n",
    "    condition: The temperature shift conditions to include (e.g, ['33to38', '38to33']).\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: The filtered dataset.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Create boolean masks for each filtering criterion\n",
    "\n",
    "    feature_mask  = data.columns.get_level_values('feature').isin(feature)\n",
    "    condition_mask  = data.columns.get_level_values('condition').isin(condition)\n",
    "    replicate_mask  = data.columns.get_level_values('replicate').isin(replicate)\n",
    "    time_mask = data.columns.get_level_values('time').isin(time)\n",
    "    feeding_mask = data.columns.get_level_values('feeding').isin(feeding)\n",
    "    # Use the masks to filter the data\n",
    "\n",
    "    #filtered_data = data.iloc[:, feature_mask & condition_mask & replicate_mask & time_mask]\n",
    "    filtered_data = data.iloc[:, feature_mask & condition_mask & replicate_mask & time_mask & feeding_mask]\n",
    "    return filtered_data\n",
    "\n",
    "def run_PCA(data, n_components=5, color_by='replicate', symbol_by='time', scale_data=True, log_transform=True):\n",
    "    \n",
    "    \"\"\"\n",
    "    Runs PCA on the provided data and generates a scatter plot of pairs of principal components.\n",
    "    It also returns a DataFrame containing the PCA loadings.\n",
    "\n",
    "    Parameters:\n",
    "    data: The gene expression data to analyze.\n",
    "    n_components: The number of principal components to calculate.\n",
    "    color_by: The metadata category to color the data points by.\n",
    "    symbol_by: The metadata category to symbolize the data points by.\n",
    "    scale_data: Whether to scale the data before PCA.\n",
    "    log_transform: Whether to apply log transformation to the data.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: A DataFrame containing the PCA loadings.\n",
    "    \"\"\"\n",
    "\n",
    "    data_tmp = data.copy()\n",
    "    \n",
    "    # Scale and/or log-transform the data if specified\n",
    "    if scale_data:\n",
    "        data = scale(data)\n",
    "    else:\n",
    "        data = data\n",
    "        \n",
    "    if log_transform:\n",
    "        data = np.log2(data + 1)\n",
    "        \n",
    "    # Fit the PCA model\n",
    "    pca = PCA(n_components=n_components)\n",
    "    pca.fit(data.T)\n",
    "\n",
    "    # Transform the data\n",
    "    pca_transformed = pca.transform(data.T)\n",
    "\n",
    "    # Create a DataFrame for the PCA results\n",
    "    labels = [f'PC{i+1}' for i in range(n_components)]\n",
    "    pca_df = pd.DataFrame(pca_transformed, columns=labels)\n",
    "    pca_df[color_by] = data_tmp.columns.get_level_values(color_by)\n",
    "    pca_df[symbol_by] = data_tmp.columns.get_level_values(symbol_by)\n",
    "\n",
    "    # Plotting each pair of consecutive principal components\n",
    "    fig, axes = plt.subplots(1, n_components -1, figsize=(30, 5))\n",
    "    for i in range(n_components-1):\n",
    "        sns.scatterplot(ax=axes[i], data=pca_df, x=f'PC{i+1}', y=f'PC{i+2}', \n",
    "        hue=color_by, style=symbol_by, s=200, palette='Set1')\n",
    "\n",
    "        axes[i].set_xlabel(f'PC{i+1} ({pca.explained_variance_ratio_[i]*100:.2f}%)')\n",
    "        axes[i].set_ylabel(f'PC{i+2} ({pca.explained_variance_ratio_[i+1]*100:.2f}%)')\n",
    "        # remove legend if it's not the first plot\n",
    "        if i != 0:\n",
    "            axes[i].get_legend().remove()\n",
    "    \n",
    "    fig.suptitle(f'PCA of Gene Expression Data (color={color_by}, symbol={symbol_by})', fontsize=16)\n",
    "\n",
    "    # Return the PCA loadings\n",
    "    df_pca = pd.DataFrame(pca.components_, columns=data_tmp.index, index=labels)\n",
    "\n",
    "    return df_pca\n",
    "\n",
    "\n",
    "## Plot Gene profile across all conditions\n",
    "def plot_gene(data, gene , xx_33_38=None, xx_38_33=None, t_2=None):\n",
    "\n",
    "    \"\"\"\n",
    "    Plots the gene expression profile across different conditions and time points.\n",
    "    \n",
    "    Parameters:\n",
    "    data: The gene expression data.\n",
    "    gene: The gene to plot.\n",
    "    xx_33_38: The simulated solution for the 33to38 condition.\n",
    "    xx_38_33: The simulated solution for the 38to33 condition.\n",
    "\n",
    "    Returns:\n",
    "    None\n",
    "    \"\"\"\n",
    "\n",
    "    #subselect\n",
    "    time= ['0', '1' ,'3' ,'6', '9']\n",
    "    replicate = ['2', '13', '11', '5']\n",
    "    dat_1= subselect(data, ['intron'], replicate, time, ['33to38'])\n",
    "    dat_2= subselect(data, ['exon'], replicate, time, ['33to38'])\n",
    "    dat_3= subselect(data, ['intron'], replicate, time, ['38to33'])\n",
    "    dat_4= subselect(data, ['exon'], replicate, time, ['38to33'])\n",
    "\n",
    "    #plot\n",
    "    t=np.array([0, 0, 1, 1, 3, 3, 6, 6, 9 , 9])\n",
    "    fig, axs = plt.subplots(2, 2, figsize=(8,8), sharex=True)\n",
    "    \n",
    "    axs[0, 0].plot(t, dat_1[dat_1.index.str.endswith(gene)].values.T, 'o', color = 'lightblue', label='intron 33to38')\n",
    "    axs[0, 0].set_title(\"intron 33to38\")\n",
    "    \n",
    "    axs[1, 0].plot(t, dat_2[dat_2.index.str.endswith(gene)].values.T, 'o', color = 'lightblue', label='exon 33to38')\n",
    "    axs[1, 0].set_title(\"exons 33to38\")\n",
    "    \n",
    "    axs[0, 1].plot(t, dat_3[dat_3.index.str.endswith(gene)].values.T, 'o', color = 'lightblue', label='intron 38to33')\n",
    "    axs[0, 1].set_title(\"intron 38to33\")\n",
    "\n",
    "    axs[1, 1].plot(t, dat_4[dat_4.index.str.endswith(gene)].values.T, 'o', color = 'lightblue', label='exon 38to33')\n",
    "    axs[1, 1].set_title(\"exon 38to33\")\n",
    "\n",
    "    if xx_33_38 is not None:\n",
    "        axs[0,0].plot(t_2, xx_33_38[:,0],label = 'P(t)')\n",
    "        axs[1,0].plot(t_2, xx_33_38[:,1])\n",
    "        axs[0,1].plot(t_2, xx_38_33[:,0])\n",
    "        axs[1,1].plot(t_2, xx_38_33[:,1])\n",
    "    \n",
    "        \n",
    "    for ax in axs.flat:\n",
    "        ax.set(xlabel='Time [h]', ylabel='RPKM')\n",
    "\n",
    "    fig.suptitle(f'Gene Expression Profile for {gene}', fontsize=16)\n",
    "    plt.show()\n",
    "\n",
    "def plot_DE(dat_all, condition_x1, condition_x2, condition_y1, condition_y2, time_x1, time_x2, time_y1, time_y2, feature_x1, feature_x2, feature_y1, feature_y2, FC=2, xlab=\"\", ylab=\"\", FC_type=\"delta\"):\n",
    "    \n",
    "    \"\"\"\n",
    "    Plots differentially expressed genes based on log fold changes between specified conditions, times, and features.\n",
    "\n",
    "    Parameters:\n",
    "    dat_all: The complete dataset with gene expression values.\n",
    "    condition_x1, condition_x2, condition_y1, condition_y2: Lists of conditions to compare.\n",
    "    time_x1, time_x2, time_y1, time_y2: Lists of time points to compare.\n",
    "    feature_x1, feature_x2, feature_y1, feature_y2: Lists of features to compare.\n",
    "    FC: Absolute fold change threshold for considering a gene differentially expressed in at least one of the two axes.\n",
    "    xlab: Label for the x-axis representing the log fold change between condition_x2 and condition_x1.\n",
    "    ylab: Label for the y-axis representing the log fold change between condition_y2 and condition_y1.\n",
    "\n",
    "    Returns:\n",
    "    Array: Boolean array indicating differentially expressed genes that meet the fold change threshold.\n",
    "    \"\"\"\n",
    "     \n",
    "    replicate = ['2', '13', '11', '5']\n",
    "\n",
    "    dat_x1=subselect(dat_all, feature_x1, replicate, time_x1, condition_x1).mean(axis=1)\n",
    "    dat_x2=subselect(dat_all, feature_x2, replicate, time_x2, condition_x2).mean(axis=1)\n",
    "\n",
    "    dat_y1=subselect(dat_all, feature_y1, replicate, time_y1, condition_y1).mean(axis=1)\n",
    "    dat_y2=subselect(dat_all, feature_y2, replicate, time_y2, condition_y2).mean(axis=1)\n",
    "\n",
    "    x=np.log2(1 + dat_x2)- np.log2(1 + dat_x1)\n",
    "    y=np.log2(1 + dat_y2) - np.log2(1 + dat_y1)\n",
    "\n",
    "    if FC_type == \"delta\":\n",
    "        pos =((x > FC) & (y > FC)) | ((x < -FC) & (y < -FC))\n",
    "\n",
    "    if FC_type == \"delta_delta\":\n",
    "        pos = abs(x-y) > FC\n",
    "\n",
    "    x_2 = x[pos]\n",
    "    y_2 = y[pos]\n",
    "    fig, ax = plt.subplots(figsize=(10,10))\n",
    "\n",
    "    ax.plot(x,y, 'o', color = 'lightblue', markersize=2)\n",
    "    ax.axhline(y=0, color='b')\n",
    "    ax.axvline(x=0, color='b')\n",
    "    ax.set_xlim(-8,8)\n",
    "    ax.set_ylim(-8,8)\n",
    "    ax.set_xlabel(xlab)\n",
    "    ax.set_ylabel(ylab)\n",
    "    texts = [ax.text(x_2[k], y_2[k], v.split(\"|\")[1]) for k, v in enumerate(dat_all.index[pos])]\n",
    "    adjust_text(texts)\n",
    "    \n",
    "    return(pos)    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Question 1\n",
    "1. Modify the code from Week 9 exercises to perform PCA using the 14 following genes (core clock genes and selected clock output genes):  \n",
    " ``['Arntl', 'Npas2', 'Clock', 'Cry1', 'Cry2', 'Per1', 'Per2', 'Per3', 'Nr1d1', 'Nr1d2', 'Rora', 'Rorc', 'Tef', 'Dbp']``.  \n",
    " First start with only the exons, then only the introns, and eventually both. Describe your observations.  \n",
    "\n",
    "2. Use the *return_amp_phase_pv* function on the $log_2$ data to assess and characterize rhythmicity genome-wide.\n",
    "    * Select genes based on p-value ($p<0.01) and amplitude (log2 fold-change>0.5) thresholds for further analysis.\n",
    "    * Provide histograms of the peak times, and amplitudes ($log_2$ peak-to-trough) for the selected genes.\n",
    "    * Perform this analysis separately for the exons and introns.\n",
    "    * Examine whether you observe patterns typically referred to as the 'morning' and 'evening' waves of gene expression.\n",
    "    \n",
    "3. Determine which genes show the highest amplitude rhythmic patterns. Are there any common themes or characteristics among these genes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>condition</th>\n",
       "      <th colspan=\"21\" halign=\"left\">WT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feeding</th>\n",
       "      <th colspan=\"21\" halign=\"left\">AL</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature</th>\n",
       "      <th colspan=\"10\" halign=\"left\">Intron</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"10\" halign=\"left\">RFP</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>time</th>\n",
       "      <th>00</th>\n",
       "      <th>02</th>\n",
       "      <th>04</th>\n",
       "      <th>06</th>\n",
       "      <th>08</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>14</th>\n",
       "      <th>16</th>\n",
       "      <th>18</th>\n",
       "      <th>...</th>\n",
       "      <th>04</th>\n",
       "      <th>06</th>\n",
       "      <th>08</th>\n",
       "      <th>10</th>\n",
       "      <th>12</th>\n",
       "      <th>14</th>\n",
       "      <th>16</th>\n",
       "      <th>18</th>\n",
       "      <th>20</th>\n",
       "      <th>22</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>replicate</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>A</th>\n",
       "      <th>...</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "      <th>D</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Arntl</th>\n",
       "      <td>-1.669041</td>\n",
       "      <td>-1.694083</td>\n",
       "      <td>-2.706959</td>\n",
       "      <td>-3.508204</td>\n",
       "      <td>-3.848818</td>\n",
       "      <td>-4.003520</td>\n",
       "      <td>-3.509194</td>\n",
       "      <td>-2.494314</td>\n",
       "      <td>-2.238684</td>\n",
       "      <td>-2.282043</td>\n",
       "      <td>...</td>\n",
       "      <td>3.022983</td>\n",
       "      <td>1.673178</td>\n",
       "      <td>-0.323042</td>\n",
       "      <td>-1.531355</td>\n",
       "      <td>0.016196</td>\n",
       "      <td>1.502431</td>\n",
       "      <td>0.612706</td>\n",
       "      <td>3.621233</td>\n",
       "      <td>2.906462</td>\n",
       "      <td>3.175841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Npas2</th>\n",
       "      <td>-1.962732</td>\n",
       "      <td>-2.192843</td>\n",
       "      <td>-3.944118</td>\n",
       "      <td>-4.209720</td>\n",
       "      <td>-4.256702</td>\n",
       "      <td>-4.273067</td>\n",
       "      <td>-4.269427</td>\n",
       "      <td>-4.246326</td>\n",
       "      <td>-4.237317</td>\n",
       "      <td>-4.149953</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.111984</td>\n",
       "      <td>-3.286106</td>\n",
       "      <td>-3.129760</td>\n",
       "      <td>-3.995698</td>\n",
       "      <td>-4.009316</td>\n",
       "      <td>-4.020822</td>\n",
       "      <td>-3.073643</td>\n",
       "      <td>-0.001143</td>\n",
       "      <td>-2.603965</td>\n",
       "      <td>-0.873309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Clock</th>\n",
       "      <td>-1.597633</td>\n",
       "      <td>-1.549136</td>\n",
       "      <td>-1.794397</td>\n",
       "      <td>-1.740208</td>\n",
       "      <td>-2.561197</td>\n",
       "      <td>-2.653104</td>\n",
       "      <td>-2.276997</td>\n",
       "      <td>-2.292100</td>\n",
       "      <td>-1.115022</td>\n",
       "      <td>-1.490327</td>\n",
       "      <td>...</td>\n",
       "      <td>1.518437</td>\n",
       "      <td>1.848053</td>\n",
       "      <td>1.144557</td>\n",
       "      <td>0.770371</td>\n",
       "      <td>1.090088</td>\n",
       "      <td>1.398042</td>\n",
       "      <td>1.619607</td>\n",
       "      <td>2.039359</td>\n",
       "      <td>1.611866</td>\n",
       "      <td>1.753404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cry1</th>\n",
       "      <td>-3.015745</td>\n",
       "      <td>-2.888921</td>\n",
       "      <td>-3.226192</td>\n",
       "      <td>-3.355850</td>\n",
       "      <td>-3.444067</td>\n",
       "      <td>-2.907677</td>\n",
       "      <td>-2.312536</td>\n",
       "      <td>-1.554579</td>\n",
       "      <td>-1.367280</td>\n",
       "      <td>-1.529984</td>\n",
       "      <td>...</td>\n",
       "      <td>1.572807</td>\n",
       "      <td>1.120746</td>\n",
       "      <td>0.413170</td>\n",
       "      <td>1.039117</td>\n",
       "      <td>1.978841</td>\n",
       "      <td>2.477187</td>\n",
       "      <td>0.829508</td>\n",
       "      <td>2.543046</td>\n",
       "      <td>3.048787</td>\n",
       "      <td>3.098065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cry2</th>\n",
       "      <td>-2.125658</td>\n",
       "      <td>-1.418011</td>\n",
       "      <td>-1.267150</td>\n",
       "      <td>-0.570111</td>\n",
       "      <td>-0.933872</td>\n",
       "      <td>-0.568488</td>\n",
       "      <td>-0.830493</td>\n",
       "      <td>-0.808692</td>\n",
       "      <td>-0.764942</td>\n",
       "      <td>-0.753802</td>\n",
       "      <td>...</td>\n",
       "      <td>2.403922</td>\n",
       "      <td>3.050125</td>\n",
       "      <td>3.369495</td>\n",
       "      <td>3.397454</td>\n",
       "      <td>3.530218</td>\n",
       "      <td>3.475627</td>\n",
       "      <td>3.426552</td>\n",
       "      <td>2.585690</td>\n",
       "      <td>3.121378</td>\n",
       "      <td>3.092411</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 144 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "condition          WT                                                    \\\n",
       "feeding            AL                                                     \n",
       "feature        Intron                                                     \n",
       "time               00        02        04        06        08        10   \n",
       "replicate           A         A         A         A         A         A   \n",
       "Gene_Symbol                                                               \n",
       "Arntl       -1.669041 -1.694083 -2.706959 -3.508204 -3.848818 -4.003520   \n",
       "Npas2       -1.962732 -2.192843 -3.944118 -4.209720 -4.256702 -4.273067   \n",
       "Clock       -1.597633 -1.549136 -1.794397 -1.740208 -2.561197 -2.653104   \n",
       "Cry1        -3.015745 -2.888921 -3.226192 -3.355850 -3.444067 -2.907677   \n",
       "Cry2        -2.125658 -1.418011 -1.267150 -0.570111 -0.933872 -0.568488   \n",
       "\n",
       "condition                                            ...                      \\\n",
       "feeding                                              ...                       \n",
       "feature                                              ...       RFP             \n",
       "time               12        14        16        18  ...        04        06   \n",
       "replicate           A         A         A         A  ...         D         D   \n",
       "Gene_Symbol                                          ...                       \n",
       "Arntl       -3.509194 -2.494314 -2.238684 -2.282043  ...  3.022983  1.673178   \n",
       "Npas2       -4.269427 -4.246326 -4.237317 -4.149953  ... -0.111984 -3.286106   \n",
       "Clock       -2.276997 -2.292100 -1.115022 -1.490327  ...  1.518437  1.848053   \n",
       "Cry1        -2.312536 -1.554579 -1.367280 -1.529984  ...  1.572807  1.120746   \n",
       "Cry2        -0.830493 -0.808692 -0.764942 -0.753802  ...  2.403922  3.050125   \n",
       "\n",
       "condition                                                                \\\n",
       "feeding                                                                   \n",
       "feature                                                                   \n",
       "time               08        10        12        14        16        18   \n",
       "replicate           D         D         D         D         D         D   \n",
       "Gene_Symbol                                                               \n",
       "Arntl       -0.323042 -1.531355  0.016196  1.502431  0.612706  3.621233   \n",
       "Npas2       -3.129760 -3.995698 -4.009316 -4.020822 -3.073643 -0.001143   \n",
       "Clock        1.144557  0.770371  1.090088  1.398042  1.619607  2.039359   \n",
       "Cry1         0.413170  1.039117  1.978841  2.477187  0.829508  2.543046   \n",
       "Cry2         3.369495  3.397454  3.530218  3.475627  3.426552  2.585690   \n",
       "\n",
       "condition                        \n",
       "feeding                          \n",
       "feature                          \n",
       "time               20        22  \n",
       "replicate           D         D  \n",
       "Gene_Symbol                      \n",
       "Arntl        2.906462  3.175841  \n",
       "Npas2       -2.603965 -0.873309  \n",
       "Clock        1.611866  1.753404  \n",
       "Cry1         3.048787  3.098065  \n",
       "Cry2         3.121378  3.092411  \n",
       "\n",
       "[5 rows x 144 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dat = pd.read_csv(\"./GSE73554_WT_AL_Intron_Exon_RFP.txt\",sep='\\t')\n",
    "dat.index = dat['Gene_Symbol']\n",
    "dat = dat.drop(['Gene_Symbol','Gene_Ensembl'], axis=1)\n",
    "dat.columns = dat.columns.str.split('_', expand = True)\n",
    "dat.columns.names = ['condition','feeding','feature','time','replicate']\n",
    "list_of_genes = ['Arntl', 'Npas2', 'Clock', 'Cry1', 'Cry2', 'Per1', 'Per2', 'Per3', 'Nr1d1', 'Nr1d2', 'Rora', 'Rorc', 'Tef', 'Dbp']\n",
    "# show the different variables in condition of dat_genes\n",
    "dat_genes.columns.get_level_values('time').unique()\n",
    "\n",
    "#extract from dat the genes in list_of_genes  \n",
    "dat_genes = dat.loc[list_of_genes]\n",
    "dat_genes.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "time= ['00', '02', '04', '06', '08', '10', '12', '14', '16', '18', '20', '22']\n",
    "replicate = ['A', 'B', 'C', 'D']\n",
    "feeding = ['AL']\n",
    "conditions = ['WT']\n",
    "feature = ['Intron', 'Exon']\n",
    "\n",
    "\n",
    "dat_intron = subselect(dat_genes, ['Intron'], replicate, time, conditions, feeding)\n",
    "dat_exon = subselect(dat_genes, ['Exon'], replicate, time, conditions, feeding)\n",
    "dat_both = subselect(dat_genes, feature, replicate, time, conditions, feeding)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\morri\\AppData\\Local\\Temp\\ipykernel_9640\\2838016254.py:58: RuntimeWarning: invalid value encountered in log2\n",
      "  data = np.log2(data + 1)\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Input X contains NaN.\nPCA does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See https://scikit-learn.org/stable/modules/impute.html You can find a list of all estimators that handle NaN values at the following page: https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\morri\\git\\Genome-to-Function-class-project3\\Exercise_Week11_12_B_final_FN.ipynb Cell 19\u001b[0m line \u001b[0;36m1\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/morri/git/Genome-to-Function-class-project3/Exercise_Week11_12_B_final_FN.ipynb#X41sZmlsZQ%3D%3D?line=6'>7</a>\u001b[0m \u001b[39m#Subselect the samples \u001b[39;00m\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/morri/git/Genome-to-Function-class-project3/Exercise_Week11_12_B_final_FN.ipynb#X41sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m dat_sub \u001b[39m=\u001b[39m subselect(dat_both, feature, replicate, time, conditions, feeding)\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/morri/git/Genome-to-Function-class-project3/Exercise_Week11_12_B_final_FN.ipynb#X41sZmlsZQ%3D%3D?line=9'>10</a>\u001b[0m df_pca \u001b[39m=\u001b[39m run_PCA(dat_both, n_components\u001b[39m=\u001b[39;49m\u001b[39m2\u001b[39;49m, color_by\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39mfeature\u001b[39;49m\u001b[39m'\u001b[39;49m, symbol_by\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39mreplicate\u001b[39;49m\u001b[39m'\u001b[39;49m, scale_data\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m, log_transform\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/morri/git/Genome-to-Function-class-project3/Exercise_Week11_12_B_final_FN.ipynb#X41sZmlsZQ%3D%3D?line=10'>11</a>\u001b[0m \u001b[39m#df_pca = run_PCA(dat_genes, n_components=2, color_by='exon', symbol_by='time', scale_data=True, log_transform=True)\u001b[39;00m\n",
      "\u001b[1;32mc:\\Users\\morri\\git\\Genome-to-Function-class-project3\\Exercise_Week11_12_B_final_FN.ipynb Cell 19\u001b[0m line \u001b[0;36m6\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/morri/git/Genome-to-Function-class-project3/Exercise_Week11_12_B_final_FN.ipynb#X41sZmlsZQ%3D%3D?line=59'>60</a>\u001b[0m \u001b[39m# Fit the PCA model\u001b[39;00m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/morri/git/Genome-to-Function-class-project3/Exercise_Week11_12_B_final_FN.ipynb#X41sZmlsZQ%3D%3D?line=60'>61</a>\u001b[0m pca \u001b[39m=\u001b[39m PCA(n_components\u001b[39m=\u001b[39mn_components)\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/morri/git/Genome-to-Function-class-project3/Exercise_Week11_12_B_final_FN.ipynb#X41sZmlsZQ%3D%3D?line=61'>62</a>\u001b[0m pca\u001b[39m.\u001b[39;49mfit(data\u001b[39m.\u001b[39;49mT)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/morri/git/Genome-to-Function-class-project3/Exercise_Week11_12_B_final_FN.ipynb#X41sZmlsZQ%3D%3D?line=63'>64</a>\u001b[0m \u001b[39m# Transform the data\u001b[39;00m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/morri/git/Genome-to-Function-class-project3/Exercise_Week11_12_B_final_FN.ipynb#X41sZmlsZQ%3D%3D?line=64'>65</a>\u001b[0m pca_transformed \u001b[39m=\u001b[39m pca\u001b[39m.\u001b[39mtransform(data\u001b[39m.\u001b[39mT)\n",
      "File \u001b[1;32mc:\\Users\\morri\\anaconda3\\envs\\Topics\\lib\\site-packages\\sklearn\\base.py:1152\u001b[0m, in \u001b[0;36m_fit_context.<locals>.decorator.<locals>.wrapper\u001b[1;34m(estimator, *args, **kwargs)\u001b[0m\n\u001b[0;32m   1145\u001b[0m     estimator\u001b[39m.\u001b[39m_validate_params()\n\u001b[0;32m   1147\u001b[0m \u001b[39mwith\u001b[39;00m config_context(\n\u001b[0;32m   1148\u001b[0m     skip_parameter_validation\u001b[39m=\u001b[39m(\n\u001b[0;32m   1149\u001b[0m         prefer_skip_nested_validation \u001b[39mor\u001b[39;00m global_skip_validation\n\u001b[0;32m   1150\u001b[0m     )\n\u001b[0;32m   1151\u001b[0m ):\n\u001b[1;32m-> 1152\u001b[0m     \u001b[39mreturn\u001b[39;00m fit_method(estimator, \u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n",
      "File \u001b[1;32mc:\\Users\\morri\\anaconda3\\envs\\Topics\\lib\\site-packages\\sklearn\\decomposition\\_pca.py:434\u001b[0m, in \u001b[0;36mPCA.fit\u001b[1;34m(self, X, y)\u001b[0m\n\u001b[0;32m    416\u001b[0m \u001b[39m@_fit_context\u001b[39m(prefer_skip_nested_validation\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n\u001b[0;32m    417\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mfit\u001b[39m(\u001b[39mself\u001b[39m, X, y\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m):\n\u001b[0;32m    418\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"Fit the model with X.\u001b[39;00m\n\u001b[0;32m    419\u001b[0m \n\u001b[0;32m    420\u001b[0m \u001b[39m    Parameters\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    432\u001b[0m \u001b[39m        Returns the instance itself.\u001b[39;00m\n\u001b[0;32m    433\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[1;32m--> 434\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_fit(X)\n\u001b[0;32m    435\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\n",
      "File \u001b[1;32mc:\\Users\\morri\\anaconda3\\envs\\Topics\\lib\\site-packages\\sklearn\\decomposition\\_pca.py:483\u001b[0m, in \u001b[0;36mPCA._fit\u001b[1;34m(self, X)\u001b[0m\n\u001b[0;32m    477\u001b[0m \u001b[39mif\u001b[39;00m issparse(X):\n\u001b[0;32m    478\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mTypeError\u001b[39;00m(\n\u001b[0;32m    479\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mPCA does not support sparse input. See \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m    480\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mTruncatedSVD for a possible alternative.\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m    481\u001b[0m     )\n\u001b[1;32m--> 483\u001b[0m X \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_validate_data(\n\u001b[0;32m    484\u001b[0m     X, dtype\u001b[39m=\u001b[39;49m[np\u001b[39m.\u001b[39;49mfloat64, np\u001b[39m.\u001b[39;49mfloat32], ensure_2d\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m, copy\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcopy\n\u001b[0;32m    485\u001b[0m )\n\u001b[0;32m    487\u001b[0m \u001b[39m# Handle n_components==None\u001b[39;00m\n\u001b[0;32m    488\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mn_components \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[1;32mc:\\Users\\morri\\anaconda3\\envs\\Topics\\lib\\site-packages\\sklearn\\base.py:605\u001b[0m, in \u001b[0;36mBaseEstimator._validate_data\u001b[1;34m(self, X, y, reset, validate_separately, cast_to_ndarray, **check_params)\u001b[0m\n\u001b[0;32m    603\u001b[0m         out \u001b[39m=\u001b[39m X, y\n\u001b[0;32m    604\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39mnot\u001b[39;00m no_val_X \u001b[39mand\u001b[39;00m no_val_y:\n\u001b[1;32m--> 605\u001b[0m     out \u001b[39m=\u001b[39m check_array(X, input_name\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mX\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mcheck_params)\n\u001b[0;32m    606\u001b[0m \u001b[39melif\u001b[39;00m no_val_X \u001b[39mand\u001b[39;00m \u001b[39mnot\u001b[39;00m no_val_y:\n\u001b[0;32m    607\u001b[0m     out \u001b[39m=\u001b[39m _check_y(y, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mcheck_params)\n",
      "File \u001b[1;32mc:\\Users\\morri\\anaconda3\\envs\\Topics\\lib\\site-packages\\sklearn\\utils\\validation.py:957\u001b[0m, in \u001b[0;36mcheck_array\u001b[1;34m(array, accept_sparse, accept_large_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, estimator, input_name)\u001b[0m\n\u001b[0;32m    951\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\n\u001b[0;32m    952\u001b[0m             \u001b[39m\"\u001b[39m\u001b[39mFound array with dim \u001b[39m\u001b[39m%d\u001b[39;00m\u001b[39m. \u001b[39m\u001b[39m%s\u001b[39;00m\u001b[39m expected <= 2.\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m    953\u001b[0m             \u001b[39m%\u001b[39m (array\u001b[39m.\u001b[39mndim, estimator_name)\n\u001b[0;32m    954\u001b[0m         )\n\u001b[0;32m    956\u001b[0m     \u001b[39mif\u001b[39;00m force_all_finite:\n\u001b[1;32m--> 957\u001b[0m         _assert_all_finite(\n\u001b[0;32m    958\u001b[0m             array,\n\u001b[0;32m    959\u001b[0m             input_name\u001b[39m=\u001b[39;49minput_name,\n\u001b[0;32m    960\u001b[0m             estimator_name\u001b[39m=\u001b[39;49mestimator_name,\n\u001b[0;32m    961\u001b[0m             allow_nan\u001b[39m=\u001b[39;49mforce_all_finite \u001b[39m==\u001b[39;49m \u001b[39m\"\u001b[39;49m\u001b[39mallow-nan\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[0;32m    962\u001b[0m         )\n\u001b[0;32m    964\u001b[0m \u001b[39mif\u001b[39;00m ensure_min_samples \u001b[39m>\u001b[39m \u001b[39m0\u001b[39m:\n\u001b[0;32m    965\u001b[0m     n_samples \u001b[39m=\u001b[39m _num_samples(array)\n",
      "File \u001b[1;32mc:\\Users\\morri\\anaconda3\\envs\\Topics\\lib\\site-packages\\sklearn\\utils\\validation.py:122\u001b[0m, in \u001b[0;36m_assert_all_finite\u001b[1;34m(X, allow_nan, msg_dtype, estimator_name, input_name)\u001b[0m\n\u001b[0;32m    119\u001b[0m \u001b[39mif\u001b[39;00m first_pass_isfinite:\n\u001b[0;32m    120\u001b[0m     \u001b[39mreturn\u001b[39;00m\n\u001b[1;32m--> 122\u001b[0m _assert_all_finite_element_wise(\n\u001b[0;32m    123\u001b[0m     X,\n\u001b[0;32m    124\u001b[0m     xp\u001b[39m=\u001b[39;49mxp,\n\u001b[0;32m    125\u001b[0m     allow_nan\u001b[39m=\u001b[39;49mallow_nan,\n\u001b[0;32m    126\u001b[0m     msg_dtype\u001b[39m=\u001b[39;49mmsg_dtype,\n\u001b[0;32m    127\u001b[0m     estimator_name\u001b[39m=\u001b[39;49mestimator_name,\n\u001b[0;32m    128\u001b[0m     input_name\u001b[39m=\u001b[39;49minput_name,\n\u001b[0;32m    129\u001b[0m )\n",
      "File \u001b[1;32mc:\\Users\\morri\\anaconda3\\envs\\Topics\\lib\\site-packages\\sklearn\\utils\\validation.py:171\u001b[0m, in \u001b[0;36m_assert_all_finite_element_wise\u001b[1;34m(X, xp, allow_nan, msg_dtype, estimator_name, input_name)\u001b[0m\n\u001b[0;32m    154\u001b[0m \u001b[39mif\u001b[39;00m estimator_name \u001b[39mand\u001b[39;00m input_name \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mX\u001b[39m\u001b[39m\"\u001b[39m \u001b[39mand\u001b[39;00m has_nan_error:\n\u001b[0;32m    155\u001b[0m     \u001b[39m# Improve the error message on how to handle missing values in\u001b[39;00m\n\u001b[0;32m    156\u001b[0m     \u001b[39m# scikit-learn.\u001b[39;00m\n\u001b[0;32m    157\u001b[0m     msg_err \u001b[39m+\u001b[39m\u001b[39m=\u001b[39m (\n\u001b[0;32m    158\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m{\u001b[39;00mestimator_name\u001b[39m}\u001b[39;00m\u001b[39m does not accept missing values\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m    159\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39m encoded as NaN natively. For supervised learning, you might want\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    169\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39m#estimators-that-handle-nan-values\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m    170\u001b[0m     )\n\u001b[1;32m--> 171\u001b[0m \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(msg_err)\n",
      "\u001b[1;31mValueError\u001b[0m: Input X contains NaN.\nPCA does not accept missing values encoded as NaN natively. For supervised learning, you might want to consider sklearn.ensemble.HistGradientBoostingClassifier and Regressor which accept missing values encoded as NaNs natively. Alternatively, it is possible to preprocess the data, for instance by using an imputer transformer in a pipeline or drop samples with missing values. See https://scikit-learn.org/stable/modules/impute.html You can find a list of all estimators that handle NaN values at the following page: https://scikit-learn.org/stable/modules/impute.html#estimators-that-handle-nan-values"
     ]
    }
   ],
   "source": [
    "# Run PCA on the gene expression data\n",
    "time= ['00', '02', '04', '06', '08', '10', '12', '14', '16', '18', '20', '22']\n",
    "replicate = ['A', 'B', 'C', 'D']\n",
    "feeding = ['AL']\n",
    "conditions = ['WT']\n",
    "feature = ['Intron', 'Exon']\n",
    "#Subselect the samples \n",
    "dat_sub = subselect(dat_both, feature, replicate, time, conditions, feeding)\n",
    "\n",
    "df_pca = run_PCA(dat_both, n_components=2, color_by='feature', symbol_by='replicate', scale_data=True, log_transform=True)\n",
    "#df_pca = run_PCA(dat_genes, n_components=2, color_by='exon', symbol_by='time', scale_data=True, log_transform=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18464198",
   "metadata": {},
   "source": [
    "#### Assess rhythmicity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86c02887",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to infer the p-value, phase, amplitude and mean from a time-serie y with size N, period T and sampling Ts. \n",
    "def return_amp_phase_pv(y, Ts, T, N):\n",
    "    \n",
    "    \"\"\"\n",
    "    Perform harmonic regression using a Fourier series to infer the amplitude, phase, p-value, and mean from a time series.\n",
    "\n",
    "    Parameters:\n",
    "    y: Time series data.\n",
    "    Ts: Sampling interval.\n",
    "    T: Period of the time series.\n",
    "    N: Number of data points in the time series.\n",
    "\n",
    "    Returns:\n",
    "    tuple: A tuple containing amplitude, phase, p-value, and mean of the time series.\n",
    "\n",
    "    Notes:\n",
    "    - The function uses Fast Fourier Transform (FFT) to decompose the time series.\n",
    "    - Harmonic regression is performed to fit a cosine wave to the data.\n",
    "    - The p-value is calculated based on the fit statistics, providing an estimate of the fit quality.\n",
    "    \"\"\"\n",
    "      \n",
    "    # Time vector for the series\n",
    "    t = np.linspace(0.0, N * Ts, N)\n",
    "\n",
    "    # Fourier Transform\n",
    "    x_fft = np.fft.fft(y)\n",
    "    freq = np.fft.fftfreq(len(y), d=Ts)\n",
    "\n",
    "    # Identify the frequency closest to 1/T\n",
    "    index, = np.where(np.isclose(freq, 1/T, atol=0.005))\n",
    "\n",
    "    # Calculate amplitude and phase\n",
    "    amp = 4 / N * np.abs(x_fft[index[0]])\n",
    "    phase = T * np.arctan2(-x_fft[index[0]].imag, x_fft[index[0]].real) / (2 * np.pi)\n",
    "\n",
    "    # Mean of the time series\n",
    "    mu = 1 / N * x_fft[0].real\n",
    "    \n",
    "    # Harmonic regression to compute residuals and statistics\n",
    "    x_hat = mu + 0.5 * amp * np.cos(2 * np.pi / T * t - 2 * np.pi * phase / T)\n",
    "    res = y - x_hat\n",
    "    sig2_1 = np.var(res)\n",
    "    sig2 = np.var(y)\n",
    "    R2 = 1 - sig2_1 / sig2\n",
    "\n",
    "    # Compute p-value\n",
    "    p = 3\n",
    "    pval = 1 - beta.cdf(R2, (p - 1) / 2, (N - p) / 2)\n",
    "\n",
    "    # Adjust phase to be within [0, T]\n",
    "    if phase < 0:\n",
    "        phase += T\n",
    "        \n",
    "    return amp, phase, pval, mu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ea44eb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_data(data, feature_type, Ts, T, N):\n",
    "    \"\"\"\n",
    "    Applies the return_amp_phase_pv function to all genes for a given feature type (intron or exon).\n",
    "\n",
    "    Parameters:\n",
    "    data: The dataset containing gene expression data.\n",
    "    feature_type: Type of feature to process ('Intron' or 'Exon').\n",
    "    Ts: Sampling time.\n",
    "    T: Period.\n",
    "    N: Number of samples.\n",
    "\n",
    "    Returns:\n",
    "    dict: A dictionary containing arrays for amplitude, phase, p-value, and mean for the specified feature type.\n",
    "    \"\"\"\n",
    "\n",
    "    feature_pos = data.columns.get_level_values('feature').isin([feature_type])\n",
    "    results = {'amp': [], 'phase': [], 'pv': [], 'mu': []}\n",
    "\n",
    "    for _, row in data.iterrows():\n",
    "        a, p, pv, mu = return_amp_phase_pv(np.array(row)[feature_pos], Ts, T, N)\n",
    "        results['amp'].append(a)\n",
    "        results['phase'].append(p)\n",
    "        results['pv'].append(pv)\n",
    "        results['mu'].append(mu)\n",
    "\n",
    "    # Convert lists to numpy arrays\n",
    "    for key in results:\n",
    "        results[key] = np.array(results[key])\n",
    "\n",
    "    return results\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b6b75841",
   "metadata": {},
   "source": [
    "#### Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "477fd66e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apply the function to both intron and exon data\n",
    "Ts = 2.0\n",
    "T = 24\n",
    "N = 48\n",
    "\n",
    "results_intron = process_data(dat, 'Intron', Ts, T, N)\n",
    "results_exon = process_data(dat, 'Exon', Ts, T, N)\n",
    "\n",
    "# Select rhythmic genes based on p-value for example\n",
    "ii=np.logical_and(results_intron['pv'] < 0.05,results_exon['pv'] < 0.05)\n",
    "#... \n",
    "#...\n",
    "\n",
    "#Perform the different plots\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de1c6f26",
   "metadata": {},
   "source": [
    "### Question 2\n",
    "1. Related to Exercise 1, study the phase relationship between the pre-mRNA and mRNA for genes that exhibit rhythmic behavior at both the pre-mRNA and mRNA levels. Use appropriate cutoffs in your analysis to select a subset of genes that are rhythmic at both the pre-mRNA and mRNA levels. Aim to include several hundred genes for a robust analysis. \n",
    "\n",
    "2. Show representative profiles of genes with short or large delays.  \n",
    "\n",
    "3. Investigate (e.g. use a scatter plot) the relation between the phase difference between pre-mRNA and mRNA and the amplitude ratio (log2 fold change) of pre-mRNA and mRNA.\n",
    "    * Discuss the plot? E.g. Mark some outliers with their gene names. *Hint*: Identify genes that show exceptionally large/small phase differences or amplitude ratios compared to the general trend. \n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
